src/sage/groups/additive_abelian/additive_abelian_wrapper.py

```diff
@@ -566,3 +618,204 @@ def _rec(j, k, c):
         return x
 
     return _rec(0, max(vals), b)
+
+
+def _expand_basis_pgroup(p, alphas, vals, beta, h, rel):
+    r"""
+    Given a basis of a `p`-subgroup of a finite abelian group
+    and an element lying outside the subgroup, extend the basis
+    to the subgroup spanned jointly by the original subgroup and
+    the new element.
+
+    Used as a subroutine in :func:`basis_from_generators`.
+
+    This function modifies ``alphas`` and ``vals`` in place.
+
+    ALGORITHM: [Suth2007]_, Algorithm 9.2
+
+    INPUT:
+
+    - ``p`` -- prime integer `p`
+    - ``alphas`` -- list; basis for a `p`-subgroup of an abelian group
+    - ``vals`` -- list; valuation at `p` of the orders of the ``alphas``
+    - ``beta`` -- element of the same abelian group as the ``alphas``
+    - ``h`` -- integer; valuation at `p` of the order of ``beta``
+    - ``rel`` -- list of integers; relation on ``alphas + [beta]``
+
+    OUTPUT: basis of the subgroup generated by ``alphas + [beta]``
+
+    EXAMPLES::
+
+        sage: from sage.groups.additive_abelian.additive_abelian_wrapper import _expand_basis_pgroup
+        sage: A = AdditiveAbelianGroup([9,3])
+        sage: alphas = [A((5,2))]
+        sage: beta = A((1,0))
+        sage: vals = [2]
+        sage: rel = next([ZZ(r),ZZ(s)] for s in range(9) for r in range(9) if s > 1 and not r*alphas[0] + s*beta)
+        sage: _expand_basis_pgroup(3, alphas, vals, beta, 2, rel)
+        sage: alphas
+        [(5, 2), (6, 2)]
+        sage: vals
+        [2, 1]
+        sage: len({i*alphas[0] + j*alphas[1] for i in range(3^2) for j in range(3^1)})
+        27
+    """
+    # The given assertions should hold, but were commented out for speed.
+
+    k = len(rel)
+    if not (isinstance(alphas, list) and isinstance(vals, list)):
+        raise TypeError('alphas and vals must be lists for mutability')
+    if not len(alphas) == len(vals) == k - 1:
+        raise ValueError(f'alphas and/or vals have incorrect length')
+#    assert not sum(r*a for r,a in zip(rel, alphas+[beta]))
+#    assert all(a.order() == p**v for a,v in zip(alphas,vals))
+
+    # step 1
+    min_r = rel[-1]
+    for i in range(k-1):
+        if not rel[i]:
+            continue
+        q = rel[i].p_primary_part(p)
+        alphas[i] *= rel[i] // q
+        rel[i] = q
+        if not min_r or q < min_r:
+            min_r = q
+    if min_r <= 0:
+        raise ValueError('rel must have nonnegative entries and at least one nonzero entry')
```

Comment:
As I understand it, the only way for `min_r` to be negative is if `rel[-1]` is negative as `p_primary_part()` always returns a positive number.
```python
sage: (-1).p_primary_part(5)
1
```
As such, you aren't testing against negative entries here.

Additionally, you could simply set `min_r = float('inf')` to remove the `not min_r` check, which you then check at the end `if min_r == float('inf')` for all zero entries.

Reply:
Oops, good catch, thanks. Should be fixed now.